src/dfx/src/lib/nns/install_nns.rs

```diff
@@ -1,32 +1,661 @@
+/// on the other canisters, however in practice this is unlikely to be a huge problem in the short term.
+///
+/// # Errors
+/// - Returns an error if the local replica type in `dfx.json` is not "system".
+/// # Panics
+/// This code is not expected to panic.
+#[context("Verifying that the local replica type is 'system'.")]
+pub fn verify_local_replica_type_is_system(env: &dyn Environment) -> anyhow::Result<()> {
+    match local_replica_type(env) {
+        Ok(ReplicaSubnetType::System) => Ok(()),
+        other => Err(anyhow!("The replica subnet_type needs to be \"system\" to run NNS canisters.  Current value: {other:?}")),
+    }
+}
+
+/// Downloads a file
+#[context("Failed to download {:?} to {:?}.", source, target)]
+pub async fn download(source: &Url, target: &Path) -> anyhow::Result<()> {
+    let buffer = reqwest::get(source.clone()).await?.bytes().await?;
+    let tmp_dir = tempfile::Builder::new().tempdir()?;
+    let downloaded_filename = {
+        let filename = tmp_dir.path().join("wasm");
+        let mut file = fs::File::create(&filename)?;
+        file.write_all(&buffer)?;
+        filename
+    };
+    fs::rename(downloaded_filename, target)?;
+    Ok(())
+}
+
+/// Downloads and unzips a file
+#[context("Failed to download and unzip {:?} from {:?}.", target, source.as_str())]
+pub async fn download_gz(source: &Url, target: &Path) -> anyhow::Result<()> {
+    let response = reqwest::get(source.clone()).await?.bytes().await?;
+    let mut decoder = Decoder::new(&response[..])?;
+    let mut buffer = Vec::new();
+    decoder.read_to_end(&mut buffer)?;
+
+    let tmp_dir = tempfile::Builder::new().tempdir()?;
+    let downloaded_filename = {
+        let filename = tmp_dir.path().join("wasm");
+        let mut file = fs::File::create(&filename)?;
+        file.write_all(&buffer)?;
+        filename
+    };
+    fs::rename(downloaded_filename, target)?;
+    Ok(())
+}
 
+/// Downloads wasm file from the main IC repo CI.
+#[context("Failed to download {} from the IC CI.", wasm_name)]
+pub async fn download_ic_repo_wasm(
+    wasm_name: &str,
+    ic_commit: &str,
+    wasm_dir: &Path,
+) -> anyhow::Result<()> {
+    fs::create_dir_all(wasm_dir)?;
+    let final_path = wasm_dir.join(&wasm_name);
+    if final_path.exists() {
+        return Ok(());
+    }
+
+    let url_str =
+        format!("https://download.dfinity.systems/ic/{ic_commit}/canisters/{wasm_name}.gz");
+    let url = Url::parse(&url_str)?;
+    println!(
+        "Downloading {}\n  from {}",
+        final_path.to_string_lossy(),
+        url_str
+    );
+    download_gz(&url, &final_path).await
+}
+
+/// Downloads all the core NNS wasms, excluding only the front-end wasms II and NNS-dapp.
+pub async fn download_nns_wasms(env: &dyn Environment) -> anyhow::Result<()> {
+    // TODO: Include the canister ID in the path.  .dfx/local/wasms/nns/${COMMIT}/....
+    let ic_commit = "3982db093a87e90cbe0595877a4110e4f37ac740"; // TODO: Where should this commit come from?
```

Comment:
This can probably come from having a build.rs read a line from  https://github.com/dfinity/sdk/blob/master/scripts/dfx-asset-sources.sh, and having https://github.com/dfinity/sdk/blob/master/scripts/write-dfx-asset-sources.sh write an appropriate rev from nix/sources.json.  Or have a build.rs that reads nix/sources.json directly.

Reply:
Support incoming: https://github.com/dfinity/sdk/pull/2520/files#diff-cfd17395fe315aa9351986a62bbc490f7b6dfc5aaf51c67f61a018d89af789d2R6